modules/konduitWrapper.js

```diff
@@ -0,0 +1,96 @@
+import { registerVideoSupport } from '../src/adServerManager';
+import { targeting } from '../src/targeting';
+import { format as buildUrl } from '../src/url';
+import * as utils from '../src/utils';
+import { config } from '../src/config';
+import { getHook } from '../src/hook';
+
+const MODULE_NAME = 'Konduit';
+
+function addLogLabel(args) {
+  args = [].slice.call(args);
+  args.unshift(`${MODULE_NAME}: `);
+  return args;
+}
+
+export function logInfo() {
+  utils.logInfo(...addLogLabel(arguments));
+}
+
+export function logError() {
+  utils.logError(...addLogLabel(arguments));
+}
+
+export function buildVastUrl(options) {
+  if (!options.params || !options.params.konduit_id) {
+    logError(`'konduit_id' parameter is required for $$PREBID_GLOBAL$$.adServers.konduit.buildVastUrl function`);
+
+    return null;
+  }
+
+  const bid = options.bid || targeting.getWinningBids()[0];
+
+  if (!bid) {
+    logError('Bid is not provided or not found');
+
+    return null;
+  }
+
+  logInfo('The following bid will be wrapped: ', bid);
+
+  const queryParams = {};
+
+  const vastUrl = obtainVastUrl(bid);
+
+  if (vastUrl) {
+    queryParams.konduit_id = options.params.konduit_id;
+    queryParams.konduit_header_bidding = 1;
+    queryParams.konduit_url = vastUrl;
+  } else {
+    logError('No VAST url found in the bid');
+  }
+
+  let resultingUrl = null;
+
+  if (queryParams.konduit_url) {
+    resultingUrl = buildUrl({
+      protocol: 'https',
+      host: 'p.konduit.me',
+      pathname: '/api/vastProxy',
+      search: queryParams
+    });
+
+    logInfo(`Konduit wrapped VAST url: ${resultingUrl}`);
+  }
+
+  return resultingUrl;
+}
+
+export function notifyTranslationModule(fn) {
+  fn.call(this, 'konduit');
+}
+
+getHook('registerAdserver').before(notifyTranslationModule);
+
+function obtainVastUrl(bid) {
+  const cacheUrl = config.getConfig('cache.url');
+  if (cacheUrl) {
+    const composedCacheUrl = `${cacheUrl}?uuid=${bid.videoCacheKey}`;
```

Comment:
@konduit-dev Hi this is a very nice looking setup. Very Clean.

But I did find one issue:

If it is a video bid, then there should always be a `vastUrl` on the bid.

Because before adding the bid to the bidsRecieved pool, if the entire VastXML is sent back, then the caching happens client side as long as `cache.url` is present, otherwise bid is thrown out.

So by the time it comes back to your module, presumably in the auction callback, the bid should have a vastUrl otherwise it is not a video bid!

Is there some other reason?

So basically you are setting the vastUrl manually if a cache.url is defined every time.

This is an issue if one of your bidders is a SERVER side cachers and returns their own VAST URL.

But the other bidders are Client side Cachers, so cache.url IS defined.

For example, Rubicon caches their Vast XML's Server Side and only returns the vastUrl,

So if you had the following:
1. Appnexus Cache as your client side config:
```javascript
pbjs.setConfig({
                cache: {
                    url: "https://prebid.adnxs.com/pbc/v1/cache"
                },
});
```

2. Rubicon And Appnexus as Video Bidders

3. Rubicon winds the bid they will return something like so:

![image](https://user-images.githubusercontent.com/13972794/65523468-2a29e380-dea1-11e9-93ff-24879e890ae7.png)

As you can see from that `bidResponse`, Rubicon had set its own `VastUrl` from its server, which then Prebid Core will not utilize the client side caching.

But your module will incorrectly generate this winning VastXML as:

https://prebid.adnxs.com/pbc/v1/cache?uuid=7ced6033-6a17-4f8b-8dde-ed769a019ed2&konduit_header_bidding=1

Since we would have the appnexus client side cache url in our pbjs config.


My reccomendation is you can leave the `composedCacheUrl` as a failsafe.

But to try the vastUrl on the bid first.

```javascript
function obtainVastUrl(bid) {

  if (vastUrl) {
    logInfo(`VAST url found in the bid - ${vastUrl}`);

    return encodeURIComponent(vastUrl);
  }

  const cacheUrl = config.getConfig('cache.url');
  if (cacheUrl) {
    const composedCacheUrl = `${cacheUrl}?uuid=${bid.videoCacheKey}`;

    logInfo(`VAST url is taken from cache.url: ${composedCacheUrl}`);

    return encodeURIComponent(composedCacheUrl);
  }
}
```

Reply:
Here is where you can see that Prebid will reject bids which do not have either:

1. a `vastUrl` returned back on the bid
OR
2. cache.url defined if only full `vastXml` is on the bid

https://github.com/prebid/Prebid.js/blob/master/src/video.js#L49